### ISSUE TITLE
A ConfigDict setting to infer a field type is optional if the default is None

### ISSUE THREAD

**Issue by DeadWisdom** [pydantic/pydantic#6546]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

In v1 giving a field a default of `None` implied the field was nullable (can be `None`), but in v2 this was changed.

I'm not sure why this was changed. If a field can be `None` by default, then this must mean it should be able to be `None`.

```python
class Foo(BaseModel):
    f1: str  # required, cannot be None
    f2: str | None  # required, can be None - same as str | None
    f3: str | None = None  # not required, can be None
    f4: str = 'Foobar'  # not required, but cannot be None
    f5: str = None # <-- not really possible
```

Technically, the specification of required and nullable could be separate, agreed. But I can't imagine a possible use-case of f5. Maybe I'm missing one. Perhaps you are trying to reduce user-errors. Either way, I will of course excuse you for being (ehem) pedantic here.

So I propose a setting in ConfigDict that lets me opt in to implied optional. This will also make many migrations to Pydantic V2 much easier.

```python
class Foo(BaseModel):
    model_config = ConfigDict(infer_optional=True)

    f1: str  # required, cannot be None
    f2: str | None  # required, can be None - same as str | None
    f3: str | None = None  # not required, can be None
    f4: str = 'Foobar'  # not required, but cannot be None
    f5: str = None # not required, implies can be None
```

Thank you so much for your hard work!

### Affected Components

- [X] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [X] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [ ] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by samuelcolvin** [comment]
Originally from [this tweet](https://twitter.com/deadwisdom/status/1678301757418491904).

Personally I'd be okay with this. :+1:

**Comment by davidhewitt** [comment]
I'm a little uneasy with this because it seems like a violation of the type system to assign `None` to a `str` field.

Do you want the schema to be nullable? If not, your problem may be resolved by disabling mypy's [`strict_optional`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-strict_optional) configuration flag without needing any changes in pydantic. (I haven't verified this.)

If you want the schema to be nullable (i.e. the type of `f5` should be considered `str | None`), it sounds a lot like mypy's [`implicit_optional`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-implicit_optional) configuration flag which does the same for function arguments. Maybe we should name the flag the same?

**Comment by DeadWisdom** [comment]
Interesting. I have to admit I don’t exactly understand how mypy works with pydantic. 

This is definitely a runtime issue, pydantic will not allow the field to be None and I want it to specifically make it Optional/Nullable. If I set implicit_optional in mypy, I still wouldn’t be able to do it.

It looks to me as this was done in mypy specifically because some libraries would not do the implicit optional. So this seems to be a case of mypy being more restrictive because the lax version was not widely supported. Pydantic then followed suit, but, imo the runtime checker should be more lax (accepting of input) than the type checker.

**Comment by adriangb** [comment]
You are fighting a losing battle here. The Python community, including MyPy, Pyright, etc. have decided that `foo: str = None` should not be allowed. I'll look into making this work in Pydantic but I don't think we should do it unless it's foolproof and easy to maintain.

**Comment by DeadWisdom** [comment]
Yes, but again, I think it’s probably right for the type checker to suggest that because the underlying schema system could require it. But pydantic is the underlying schema system so it shouldn’t.

Or at least let me make it optional like in mypy.

Thank you for considering it, I know you have a lot of trade offs to weigh.

**Comment by gmetzker** [comment]
This caught me by surprise as well.  After converting a large code base I still missed a few fields.

At the very least it would be nice to call it out in the Migration guide, something like:




State|Field Defintion
-- | --
Required, cannot be None | f1: str
Not required, cannot be None, is 'abc' by default | f2: str = 'abc'
Required, can be None | f3: Optional[str]
Not required, can be None, is None by default | f4: Optional[str] = None
Not required, can be None, is 'abc' by default | f5: Optional[str] = 'abc'
Required, can be any type (including None) | f6: Any
Not required, can be any type (including None) | f7: Any = None
Not required, but cannot be explicitly set to None.  No longer works in V2. Consider switch to `f8: Optional[str] = None` | f8: str = None

**Comment by Viicos** [comment]
Closing as per what was said above: all the typing ecosystem have moved away from this implicit `| None` behavior. Issues labeled with https://github.com/pydantic/pydantic/labels/topic-unset  can be followed to get updates on a way to support _non required, cannot be none_ scenarios.
